<launch>
    <arg name="input_cloud_topic" default="/points_raw" />
    <arg name="filtered_cloud_topic" default="/filter_points" />
    <arg name="ground_cloud_topic" default="/ground_points" />
    <arg name="scene_cloud_topic" default="/scene_points" />
    <arg name="vis_cluster_cloud_topic" default="cluster_points"/>
    <arg name="occ_grid_topic" default="/minibus/realtime_cost_map"/>
    <arg name="obj_array_topic" default="/zzz/detection/lidar_detector/objects"/>
    <arg name="marker_array_topic" default="/box_marker"/>
    <arg name="boundary_marker_topic" default="/boundary_marker"/>
    
    <!-- preprocessing -->
        <!-- cloud_range -->
    <!-- <arg name = "cloud_range_x_min" default = "-7"/>
    <arg name = "cloud_range_y_min" default = "-5"/>
    <arg name = "cloud_range_z_min" default = "-2.0"/>
    <arg name = "cloud_range_x_max" default = "7"/>
    <arg name = "cloud_range_y_max" default = "30.0"/>
    <arg name = "cloud_range_z_max" default = "3.0"/> -->
    <arg name = "cloud_range_x_min" default = "-30"/>
    <arg name = "cloud_range_y_min" default = "-30"/>
    <arg name = "cloud_range_z_min" default = "-2.0"/>
    <arg name = "cloud_range_x_max" default = "30"/>
    <arg name = "cloud_range_y_max" default = "30.0"/>
    <arg name = "cloud_range_z_max" default = "3.0"/>

    <arg name = "ego_size_x_min" default = "-1.05"/>
    <arg name = "ego_size_y_min" default = "-0.44"/>
    <arg name = "ego_size_z_min" default = "-2.0"/>
    <arg name = "ego_size_x_max" default = "1.05"/>
    <arg name = "ego_size_y_max" default = "0.44"/>
    <arg name = "ego_size_z_max" default = "3.0"/>

        <!-- voxel filter leaf size -->
    <arg name = "voxel_filter_leaf_size_x" default="0.05"/>
    <arg name = "voxel_filter_leaf_size_y" default="0.05"/>
    <arg name = "voxel_filter_leaf_size_z" default="0.05"/>

    <!-- ground filter -->
    <arg name = "num_ground_seeds" default = "15"/>
    <arg name = "lidar_height" default ="1.882" />
    <arg name = "th_ground_seeds" default = "1.2"/>
    <arg name = "iter_ground_plane" default = "3"/>
    <arg name = "th_ground_plane" default = "0.2"/>

    <!-- boundary make -->
    <arg name = "max_distance" default = "30"/>

    <!-- cloud cluster -->
    <arg name = "clip_height" default = "2.0"/>
        <!-- scene_split -->
    <arg name = "scene_slit_0" default = "20.0"/>
    <arg name = "scene_slit_1" default = "35.0"/>
    <arg name = "scene_slit_2" default = "50.0"/>
    <arg name = "scene_slit_3" default = "70.0"/>

        <!-- cluster_leaf_sizes -->
    <arg name = "cluster_leaf_sizes_0" default = "0.05"/>
    <arg name = "cluster_leaf_sizes_1" default = "0.05"/>
    <arg name = "cluster_leaf_sizes_2" default = "0.05"/>
    <arg name = "cluster_leaf_sizes_3" default = "0.05"/>
    <arg name = "cluster_leaf_sizes_4" default = "0.05"/>

        <!-- cluster_ths -->
    <arg name = "cluster_ths_0" default = "0.2"/>
    <arg name = "cluster_ths_1" default = "0.3"/>
    <arg name = "cluster_ths_2" default = "0.3"/>
    <arg name = "cluster_ths_3" default = "0.3"/>
    <arg name = "cluster_ths_4" default = "0.3"/>

    <arg name = "cluster_min_size" default = "10"/>
    <arg name = "cluster_max_size" default = "5000"/>
 
     <node pkg= "cluster_detector" type="cluster_detector_node" name="cluster_detector_node" output="screen">
        <param name="input_cloud_topic" value="$(arg input_cloud_topic)"/>
        <param name="filtered_cloud_topic" value="$(arg filtered_cloud_topic)"/>
        <param name="ground_cloud_topic" value="$(arg ground_cloud_topic)"/>
        <param name="scene_cloud_topic" value="$(arg scene_cloud_topic)"/>
        <param name="vis_cluster_cloud_topic" value="$(arg vis_cluster_cloud_topic)"/>
        <param name="occ_grid_topic" value="$(arg occ_grid_topic)"/>
        <param name="obj_array_topic" value="$(arg obj_array_topic)"/>
        <param name="marker_array_topic" value="$(arg marker_array_topic)"/>
        <param name="boundary_marker_topic" value="$(arg boundary_marker_topic)"/>

        
        
        <!-- preprocessing -->
            <!-- cloud_range -->
        <param name="cloud_range_x_min" value="$(arg cloud_range_x_min)" />
        <param name="cloud_range_y_min" value="$(arg cloud_range_y_min)" />
        <param name="cloud_range_z_min" value="$(arg cloud_range_z_min)" />
        <param name="cloud_range_x_max" value="$(arg cloud_range_x_max)" />
        <param name="cloud_range_y_max" value="$(arg cloud_range_y_max)" />
        <param name="cloud_range_z_max" value="$(arg cloud_range_z_max)" />

        <param name="ego_size_x_min" value="$(arg ego_size_x_min)" />
        <param name="ego_size_y_min" value="$(arg ego_size_y_min)" />
        <param name="ego_size_z_min" value="$(arg ego_size_z_min)" />
        <param name="ego_size_x_max" value="$(arg ego_size_x_max)" />
        <param name="ego_size_y_max" value="$(arg ego_size_y_max)" />
        <param name="ego_size_z_max" value="$(arg ego_size_z_max)" />




            <!-- voxel filter leaf size -->
        <param name="voxel_filter_leaf_size_x" value="$(arg voxel_filter_leaf_size_x)" />
        <param name="voxel_filter_leaf_size_y" value="$(arg voxel_filter_leaf_size_y)" />
        <param name="voxel_filter_leaf_size_z" value="$(arg voxel_filter_leaf_size_z)" />

        <!-- ground filter -->
        <param name="num_ground_seeds" value="$(arg num_ground_seeds)" />
        <param name="lidar_height" value="$(arg lidar_height)" />
        <param name="th_ground_seeds" value="$(arg th_ground_seeds)" />
        <param name="iter_ground_plane" value="$(arg iter_ground_plane)" />
        <param name="th_ground_plane" value="$(arg th_ground_plane)" />
        <!-- boundary make -->
        <param name="max_distance" value="$(arg max_distance)"/>
        <!-- cloud cluster -->
        <param name="clip_height" value="$(arg clip_height)" />
            <!-- scene_split -->
        <param name="scene_slit_0" value="$(arg scene_slit_0)" />
        <param name="scene_slit_1" value="$(arg scene_slit_1)" />
        <param name="scene_slit_2" value="$(arg scene_slit_2)" />
        <param name="scene_slit_3" value="$(arg scene_slit_3)" />

            <!-- cluster_leaf_sizes -->
        <param name="cluster_leaf_sizes_0" value="$(arg cluster_leaf_sizes_0)" />
        <param name="cluster_leaf_sizes_1" value="$(arg cluster_leaf_sizes_1)" />
        <param name="cluster_leaf_sizes_2" value="$(arg cluster_leaf_sizes_2)" />
        <param name="cluster_leaf_sizes_3" value="$(arg cluster_leaf_sizes_3)" />
        <param name="cluster_leaf_sizes_4" value="$(arg cluster_leaf_sizes_4)" />

            <!-- cluster_ths -->
        <param name="cluster_ths_0" value="$(arg cluster_ths_0)" />
        <param name="cluster_ths_1" value="$(arg cluster_ths_1)" />
        <param name="cluster_ths_2" value="$(arg cluster_ths_2)" />
        <param name="cluster_ths_3" value="$(arg cluster_ths_3)" />
        <param name="cluster_ths_4" value="$(arg cluster_ths_4)" />
        
        <param name="cluster_min_size" value="$(arg cluster_min_size)" />
        <param name="cluster_max_size" value="$(arg cluster_max_size)" />

    </node>
</launch>
